lib/c-api/src/wasm_c_api/unstable/wasi.rs

```diff
@@ -0,0 +1,199 @@
+/// implementation.
+///
+/// And that's why we have the following `cbindgen_hack`
+/// module.
+///
+/// But this module must not be compiled by `rustc`. How to force
+/// `rustc` to ignore a module? With conditional compilation. Because
+/// `cbindgen` does not support conditional compilation, it will
+/// always _ignore_ the `#[cfg]` attribute, and will always read the
+/// content of the module.
+///
+/// Sorry.
+#[doc(hidden)]
+#[cfg(__cbindgen_hack__ = "yes")]
+mod __cbindgen_hack__ {
+    use super::*;
+
+    #[repr(C)]
+    pub struct wasmer_named_extern_vec_t {
+        pub size: usize,
+        pub data: *mut *mut wasmer_named_extern_t,
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn wasmer_named_extern_vec_new(
+        out: *mut wasmer_named_extern_vec_t,
+        length: usize,
+        init: *const *mut wasmer_named_extern_t,
+    ) {
+        unimplemented!()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn wasmer_named_extern_vec_new_uninitialized(
+        out: *mut wasmer_named_extern_vec_t,
+        length: usize,
+    ) {
+        unimplemented!()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn wasmer_named_extern_vec_copy(
+        out_ptr: &mut wasmer_named_extern_vec_t,
+        in_ptr: &wasmer_named_extern_vec_t,
+    ) {
+        unimplemented!()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn wasmer_named_extern_vec_delete(
+        ptr: Option<&mut wasmer_named_extern_vec_t>,
+    ) {
+        unimplemented!()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn wasmer_named_extern_vec_new_empty(
+        out: *mut wasmer_named_extern_vec_t,
+    ) {
+        unimplemented!()
+    }
+}
+
+/// Non-standard function to get the module name of a
+/// `wasmer_named_extern_t`.
+///
+/// The returned value isn't owned by the caller.
+#[no_mangle]
+pub extern "C" fn wasmer_named_extern_module(
+    named_extern: Option<&wasmer_named_extern_t>,
+) -> Option<&wasm_name_t> {
+    Some(named_extern?.module.as_ref())
+}
+
+/// Non-standard function to get the name of a `wasmer_named_extern_t`.
+///
+/// The returned value isn't owned by the caller.
+#[no_mangle]
+pub extern "C" fn wasmer_named_extern_name(
+    named_extern: Option<&wasmer_named_extern_t>,
+) -> Option<&wasm_name_t> {
+    Some(named_extern?.name.as_ref())
+}
+
+/// Non-standard function to get the wrapped extern of a
+/// `wasmer_named_extern_t`.
+///
+/// The returned value isn't owned by the caller.
+#[no_mangle]
+pub extern "C" fn wasmer_named_extern_unwrap(
+    named_extern: Option<&wasmer_named_extern_t>,
+) -> Option<&wasm_extern_t> {
+    Some(named_extern?.r#extern.as_ref())
+}
+
+/// Non-standard function to get the imports needed for the WASI
+/// implementation with no particular order. Each import has its
+/// associated module name and name, so that it can be re-order later
+/// based on the `wasm_module_t` requirements.
+#[no_mangle]
+pub unsafe extern "C" fn wasi_get_unordered_imports(
```

Comment:
Perhaps worth considering but maybe we should put `WASI` related functions in their own header files. I'll defer to NIck's judgement about header file / C API organization but just a thought. This doesn't have to be done in this PR, but you separating out WASI into its own sub-module here made me think of that

Reply:
Since this function is unstable, I prefer to keep it here. But the function is re-exported inside the `wasi` module.

Note that it doesn't really matter since everything is flat inside the `.h` file at the end of the process.